### PR TITLE
Save while navigating - course creation

### PR DESCRIPTION
**Pull request by magnusklundgren**
## Description
This PBI reworks the way we create and update courses, by storing the data in cache and sending the entire course data to the backend at once instead of updating every section, component etc. by itself. 
We also moved the handling of images and videos to the backend, so it is uploaded from there. 

_There is a backend PR that handles all the backend logic relevant to this PR, but they can be reviewed separately._

## Changes
- Added a CourseStore & ProviderComponent that wraps CourseManager.tsx which keeps tracks of all Course, Section, Lecture, Exercise and Media state, and allows all the nested components in CourseManager to easily Read/Use, Write, Update and delete this data to allow for persistent local data managing when creating or managing courses until the final submission. 
- Added CourseStore.tsx which encapsulates the data and functions as descriped above
- Modified the CourseManager.tsx to fetch course info and image on mount and add it to the cache, if it's an existing course that is being managed. 
- Modified CourseComponent.tsx to use the CourseStore, and changing the course state on field change, buttons are now disabled and given opacity if all required fields are not filled. 
- Modified SectionCreation.tsx to use the CourseStore and more specifically load all the data to the cache after they are initially fetched; Now possible to DragNDrop Sections and Components, Now possible to update LectureVideos, Buttons are given opacity and disabled if a section contains an empty title or description. 
- Modified CreateLecturePopup, CreateExercisePopup, EditLecturePopup, EditExercisePopup;
All these files also uses the newly implemented store so data is only fetched on the initial render. 
- We updated the courseService to have 2 service functions CreateCourse and UpdateCourse,  that both utilizes our new backend route and sends an entire FormData object with CourseInfo, Sections and Files to the server for creation or updating.

- We've integrated React-testing-library, so it's now possible to test on functions that uses and changes state, which have been applied to all the tests we've written for the courseStore.  (Documentation is in the progress of being added to resources for this). 

## Related Issues

https://github.com/orgs/Educado-App/projects/14/views/12?pane=issue&itemId=82190855&issue=Educado-App%7Ceducado-frontend%7C142

## Checklist

- [x] Code has been tested locally and passes all relevant tests.
- [x] Documentation has been updated to reflect the changes, if applicable.
- [x] Code follows the established coding style and guidelines of the project.
- [x] All new and existing tests related to the changes have passed.
- [x] Any necessary dependencies or new packages have been properly documented.
- [x] Pull request title and description are clear and descriptive.
- [x] Reviewers have been assigned to the pull request.
- [x] Any potential security implications have been considered and addressed.
- [x] Performance impact of the changes has been evaluated, if relevant.
- [x] Changes have been approved by Product Owner
- [x] Code has been analyzed with CodeScene
- [x] I have reviewed the pull request myself

## If mobile/frontend pull request, what version of the backend is it stable, and running on?

Commit id can be found under branch --> commits.

Branch: deadlocked-courseLoading
commit: 9cc7733c1c9d105636b0e3407d42dc1ce713bf2b

## Link to documentation

We've written documentation for how to unit test hooks and contexts in jest with the @testing-library/react-hooks library, as we've done this for testing our CourseStore and it's exported Hooks :). 
Link to documentation PR: https://github.com/Educado-App/resources/pull/1/files

## Notes for Reviewers
We're aware that this is a large pull request most of the added code is just simple getter/setter/ or deletion functions added in courseStore which should be quite straightforward to go through, In all the other files it's mostly been about integration them with the courseStore and make sure everything worked nicely, while additionally doing alot of cleaning and refactoring. We're happy to collaborate and answer any questions in a call if you might have any.

### CODE REVIEW

**Comment by BabaWolo** [comment]
Hey it might be that I'm not on the correct branches (or something else) but I receive a 404 error trying to upload a new course with video lectures in the section.
image

I'm using the deadlock-#142 frontend branch, deadlocked-courseLoading back-end branch, and dev for transcoding.

Can you take a look to see if you get the same error, or help me out ? :D

File: app/src/components/Courses/ExploreCardPreview.tsx [hunk 1]
```diff
@@ -167,7 +162,7 @@ const ExploreCardPreview: React.FC<ExploreCardPreviewProps> = ({ course }) => {
                 <div className="h-px bg-gray-200 w-full" />
 
                 {/* Description */}
-                {course.description && (
+                {course.description !== "" && (
```
Comment (MBryder) [inline app/src/components/Courses/ExploreCardPreview.tsx:165]: Do you make sure that description is never null or undefined?

File: app/src/components/Courses/CoursePreview.tsx
Comment (MBryder) [inline app/src/components/Courses/CoursePreview.tsx:1]: The main Method is tagged with Complex Method and Large Method. is it possible to make helper functions for some of your parts?

File: app/src/components/Courses/CourseComponent.tsx
Comment (MBryder) [inline app/src/components/Courses/CourseComponent.tsx:1]: The method is marked with both Complex and Large Method is it possible to make helper functions for some of your parts?

**Comment by magnusklundgren** [comment]
> Hey it might be that I'm not on the correct branches (or something else) but I receive a 404 error trying to upload a new course with video lectures in the section. image
> 
> I'm using the deadlock-#142 frontend branch, deadlocked-courseLoading back-end branch, and dev for transcoding.
> 
> Can you take a look to see if you get the same error, or help me out ? :D
> 
> Can't create a PR for our PBI before this is fixed and I can test...
> 
> @magnusklundgren

Sorry just got time to look at this now. Maybe try to use the production transcoding service and adding the link to .env in bakcend, instead of running it locally

**Comment by Laskiri** [comment]
I'm on discord for the rest of the day if you have any inquiries

File: app/src/components/Courses/ExploreCardPreview.tsx (hunk 1 above)
Comment (Laskiri) [reply app/src/components/Courses/ExploreCardPreview.tsx:165]: Course is initialized as a new course in the store "const [course, setCourse] = useState<Course>({title: "",
    description: "",
    category: "personal finance",
    difficulty: 1,
    status: "draft",
    estimatedHours: 0,
    sections: [],
    coverImg: "",
    _id: "0",
    creator: "",
    });" unless it's an existing one that is clicked onto and then the fetched data will set the course accordingly so description will never be undefined not according to typescripts type inference, or the implementation right now. Also a course cannot be saved anymore unless all required fields are filled out, so i dont expect it to be a problem. It was mostly just typescript that was raising noise which lead to that small change when we merged with dev.

File: app/src/components/Courses/CourseComponent.tsx
Comment (Laskiri) [reply app/src/components/Courses/CourseComponent.tsx:1]: i'll look into it :)!

Comment (Laskiri) [reply app/src/components/Courses/CourseComponent.tsx:1]: I've removed a bit of the cyclic complexity in all the main 3 courseManagement components by abstracting shared logic and helper functions into a hook called useCourseManagingHelpers. The only thing that might need to be adressed is the size of the components, as there is alot of jsx in some of them but i think it's better to leave for the future as we've already done alot of changes and i think it's important to get them forward in the process soon unless any other main concerns are discovered :D

**Review by MBryder — APPROVED** [review]

File: app/src/hooks/useCourseManagingHelper.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            updatedCourse.courseInfo.status = "published";
+            const formData = prepareFormData(updatedCourse);
+
+            await submitFunction(formData, token);
+            onSuccessfulSubmit();
+        }
+        catch (err) {
+            console.error(err);
+        }
+    };
```
[the remark this answers is not in the document]
Comment (Laskiri) [reply app/src/hooks/useCourseManagingHelper.ts]: yes it is handled 👍

**Pull request merged**